File: packages/go-agentenkit/pricing/pricing_test.go
package pricing_test

import (
	"context"
	"errors"
	"testing"

	"github.com/eadwinCode/agentic-kit/packages/go-agentenkit/ports"
	"github.com/eadwinCode/agentic-kit/packages/go-agentenkit/pricing"
)

var table = pricing.Table{
	// Anthropic's published Sonnet prices, per million tokens.
	"claude-sonnet-4": {
		InputPerMillion: 3, CacheReadPerMillion: 0.30,
		CacheWritePerMillion: 3.75, OutputPerMillion: 15,
	},
	"gpt-4o-2024-11-20": {InputPerMillion: 2.5, OutputPerMillion: 10},
}

func price(t *testing.T, p ports.Pricer, u ports.NewUsage) *ports.Cost {
	t.Helper()
	c, err := p.Price(context.Background(), u)
	if err != nil {
		t.Fatalf("price: %v", err)
	}
	return c
}

func TestTable_PricesPerMillionTokensIntoMicros(t *testing.T) {
	// 1M input at $3 is $3.00, which is 3_000_000 micros.
	got := price(t, table, ports.NewUsage{Model: "claude-sonnet-4", InputTokens: 1_000_000})
	if got.Micros != 3_000_000 {
		t.Fatalf("input: got %d micros, want 3000000", got.Micros)
	}
	// A realistic mixed call: 12k fresh input, 40k cache reads, 8k cache
	// writes, 900 output.
	//   12_000×3 + 40_000×0.30 + 8_000×3.75 + 900×15 = 36_000 + 12_000 + 30_000 + 13_500
	got = price(t, table, ports.NewUsage{
		Model: "claude-sonnet-4", InputTokens: 12_000, CacheReadInputTokens: 40_000,
		CacheWriteInputTokens: 8_000, OutputTokens: 900,
	})
	if got.Micros != 91_500 {
		t.Fatalf("mixed: got %d micros, want 91500", got.Micros)
	}
	if got.Currency != "USD" || got.Source != "table" {
		t.Fatalf("got %+v", got)
	}
}

func TestTable_FallsBackToTheWireIdThenTheBaseKey(t *testing.T) {
	// The registry key is unknown but the provider reported a wire id the
	// table does know.
	if c := price(t, table, ports.NewUsage{Model: "fast", ModelID: "gpt-4o-2024-11-20", OutputTokens: 1_000_000}); c == nil || c.Micros != 10_000_000 {
		t.Fatalf("wire id fallback: %+v", c)
	}
	// A "@variant" suffix on the registry key still finds the base model.
	if c := price(t, table, ports.NewUsage{Model: "claude-sonnet-4@high", InputTokens: 1_000_000}); c == nil || c.Micros != 3_000_000 {
		t.Fatalf("variant fallback: %+v", c)
	}
	// A model nobody knows is not priced, rather than priced at zero.
	if c := price(t, table, ports.NewUsage{Model: "who-knows", InputTokens: 1_000}); c != nil {
		t.Fatalf("unknown model priced: %+v", c)
	}
}

func TestReceipt_ReadsWhatTheProviderAlreadyComputed(t *testing.T) {
	p := pricing.Receipt(func(meta map[string]any) (int64, bool) {
		headers, ok := meta["responseHeaders"].(map[string]string)
		if !ok {
			return 0, false
		}
		if headers["x-cost-micros"] == "4200" {
			return 4200, true
		}
		return 0, false
	})
	got := price(t, p, ports.NewUsage{ProviderMetadata: map[string]any{
		"responseHeaders": map[string]string{"x-cost-micros": "4200"},
	}})
	if got.Micros != 4200 || got.Source != "receipt" {
		t.Fatalf("got %+v", got)
	}
	// No receipt on this call: say nothing, so the next pricer can try.
	if c := price(t, p, ports.NewUsage{ProviderMetadata: map[string]any{}}); c != nil {
		t.Fatalf("answered without a receipt: %+v", c)
	}
}

func TestChain_TakesTheFirstAnswerAndSkipsFailures(t *testing.T) {
	boom := ports.PricerFunc(func(context.Context, ports.NewUsage) (*ports.Cost, error) {
		return nil, errors.New("price service down")
	})
	silent := ports.PricerFunc(func(context.Context, ports.NewUsage) (*ports.Cost, error) {
		return nil, nil
	})
	// A pricer that errors is skipped rather than believed, and the table
	// behind it still answers.
	got := price(t, pricing.Chain(boom, silent, table),
		ports.NewUsage{Model: "claude-sonnet-4", InputTokens: 1_000_000})
	if got.Micros != 3_000_000 {
		t.Fatalf("got %+v", got)
	}
	// Nobody could price it: the error is what comes back, so a caller can
	// see that pricing is broken rather than that the call was free.
	if _, err := pricing.Chain(boom, silent).Price(context.Background(), ports.NewUsage{}); err == nil {
		t.Fatal("want the first error back when nothing answered")
	}
	// Nothing to say and nothing wrong is not an error.
	c, err := pricing.Chain(silent).Price(context.Background(), ports.NewUsage{})
	if c != nil || err != nil {
		t.Fatalf("got %+v, %v", c, err)
	}
}

func TestMicrosRoundTrip(t *testing.T) {
	if got := pricing.Micros(0.25); got != 250_000 {
		t.Fatalf("Micros(0.25) = %d", got)
	}
	if got := pricing.Amount(250_000); got != 0.25 {
		t.Fatalf("Amount(250000) = %v", got)
	}
	if got := pricing.Format(250_000, ""); got != "0.2500 USD" {
		t.Fatalf("Format = %q", got)
	}
}

File: packages/agentenkit/test/pricing.test.ts
import { describe, expect, it } from 'bun:test';
import * as pricing from '../src/pricing.js';
import type { NewUsage } from '../src/core/types.js';
import type { Pricer } from '../src/ports/runtime.js';

// The shipped pricers (§4). The same cases run in the Go package
// (pricing/pricing_test.go).

const table = pricing.table({
  // Anthropic's published Sonnet prices, per million tokens.
  'claude-sonnet-4': {
    inputPerMillion: 3, cacheReadPerMillion: 0.3, cacheWritePerMillion: 3.75, outputPerMillion: 15,
  },
  'gpt-4o-2024-11-20': { inputPerMillion: 2.5, outputPerMillion: 10 },
});

const usage = (u: Partial<NewUsage>): NewUsage => ({
  kind: 'step', step: 1, outcome: 'finished',
  inputTokens: 0, cacheReadInputTokens: 0, cacheWriteInputTokens: 0, outputTokens: 0,
  reasoningTokens: 0, totalTokens: 0, ...u,
});
const price = (p: Pricer, u: Partial<NewUsage>) => Promise.resolve(p.price(usage(u)));

describe('table', () => {
  it('prices per million tokens into micros', async () => {
    expect((await price(table, { model: 'claude-sonnet-4', inputTokens: 1_000_000 }))?.micros).toBe(3_000_000);
    // 12_000×3 + 40_000×0.30 + 8_000×3.75 + 900×15 = 36_000 + 12_000 + 30_000 + 13_500
    const got = await price(table, {
      model: 'claude-sonnet-4', inputTokens: 12_000, cacheReadInputTokens: 40_000,
      cacheWriteInputTokens: 8_000, outputTokens: 900,
    });
    expect(got).toEqual({ micros: 91_500, currency: 'USD', source: 'table' });
  });

  it('falls back to the wire id then the base key', async () => {
    expect((await price(table, { model: 'fast', modelId: 'gpt-4o-2024-11-20', outputTokens: 1_000_000 }))?.micros)
      .toBe(10_000_000);
    expect((await price(table, { model: 'claude-sonnet-4@high', inputTokens: 1_000_000 }))?.micros).toBe(3_000_000);
    // A model nobody knows is not priced, rather than priced at zero.
    expect(await price(table, { model: 'who-knows', inputTokens: 1_000 })).toBeNull();
  });

  it('a missing cache rate is priced as input', async () => {
    // gpt-4o has no cache rates in this table: its cache reads and writes are
    // priced as input, never at 0.
    const got = await price(table, {
      model: 'gpt-4o-2024-11-20', cacheReadInputTokens: 1_000_000, cacheWriteInputTokens: 1_000_000,
    });
    expect(got?.micros).toBe(5_000_000); // 2 × $2.50
  });

  it('a model named like an object key is not found', async () => {
    for (const name of ['constructor', 'toString', '__proto__']) {
      expect(await price(table, { model: name, inputTokens: 1_000 })).toBeNull();
    }
  });
});

describe('receipt', () => {
  it('reads what the provider already computed', async () => {
    const p = pricing.receipt((meta) => {
      const headers = meta.responseHeaders as Record<string, string> | undefined;
      return headers?.['x-cost-micros'] === '4200' ? 4200 : null;
    });
    const got = await price(p, { providerMetadata: { responseHeaders: { 'x-cost-micros': '4200' } } });
    expect(got?.micros).toBe(4200);
    expect(got?.source).toBe('receipt');
    // No receipt on this call: say nothing, so the next pricer can try.
    expect(await price(p, { providerMetadata: {} })).toBeNull();
  });

  it('a negative receipt is left unpriced', async () => {
    expect(await price(pricing.receipt(() => -5), { providerMetadata: { x: 1 } })).toBeNull();
    // Nor one that is not a finite number; a fraction is rounded to whole micros.
    expect(await price(pricing.receipt(() => Number.NaN), { providerMetadata: { x: 1 } })).toBeNull();
    expect((await price(pricing.receipt(() => 41.6), { providerMetadata: { x: 1 } }))?.micros).toBe(42);
  });
});

describe('chain', () => {
  it('takes the first answer and skips failures', async () => {
    const boom: Pricer = { price: () => { throw new Error('price service down'); } };
    const silent: Pricer = { price: () => null };
    expect((await price(pricing.chain(boom, silent, table), { model: 'claude-sonnet-4', inputTokens: 1_000_000 }))?.micros)
      .toBe(3_000_000);
    // Nobody could price it: the error is what comes back.
    await expect(price(pricing.chain(boom, silent), {})).rejects.toThrow('price service down');
    // Nothing to say and nothing wrong is not an error.
    expect(await price(pricing.chain(silent), {})).toBeNull();
  });
});

describe('money helpers', () => {
  it('micros round trip', () => {
    expect(pricing.micros(0.25)).toBe(250_000);
    expect(pricing.amount(250_000)).toBe(0.25);
    expect(pricing.format(250_000)).toBe('0.2500 USD');
  });
});
